r/src/find_met_files.r

```diff
@@ -1,42 +1,50 @@
 #' find_met_files searches for meteorological data files
-#' @author Ben Fasoli
+#' @author Ben Fasoli & James Mineau
 #'
 #' Searches for available meteorological files matching the given strftime
 #' compatible file naming convention
 #'
 #' @param t_start time of simulation start
+#' @param n_hours number of hours to run each simulation; negative indicates
+#'   backward in time
+#' @param n_hours_per_met_file number of hours of meteorological data in each
+#'   met file
 #' @param met_file_format grep compatible file naming convention to identify
 #'   meteorological data files necessary for the timing of the simulation
 #'   indicated by \code{t_start} and \code{n_hours}
-#' @param n_hours number of hours to run each simulation; negative indicates
-#'   backward in time
 #' @param met_path directory to find meteorological data
 #'
 #' @import dplyr
 #' @export
 
-find_met_files <- function(t_start, met_file_format, n_hours, met_path) {
+find_met_files <- function(t_start, n_hours, n_hours_per_met_file,
+                           met_file_format, met_path) {
   require(dplyr)
-  
-  is_backward <- n_hours < 0
-  
-  # TODO: implement n_hours_per_met_file to better determine file names at
-  # varying time resolutions
-  request <- as.POSIXct(t_start, tz='UTC') %>%
-    c(. + c(1, -1, n_hours, is_backward * (n_hours - 5)) * 3600) %>%
-    range() %>%
-    (function(x) seq(x[1], x[2], by = 'hour')) %>%
-    strftime(tz = 'UTC', format = met_file_format)
-  
+
+  ts <- as.POSIXct(t_start, tz = 'UTC')
+  is_backward <- as.numeric(n_hours < 0)
+  ib <- ifelse(is_backward, 1, -1)
+
+  # Generate the list of files to search for
+  request <- seq(
+    ts - (ib * as.difftime(abs(n_hours) + ib, units = 'hours')),
+    ts + (ib * as.difftime(n_hours_per_met_file - is_backward, units = 'hours')),
+    by = ib * 3600
+  ) %>%
+    strftime(tz = 'UTC', format = met_file_format) %>%
+    unique()
+
+  # Find the available files
   available <- dir(met_path, full.names = T, recursive = T)
   available <- available[!grepl('.lock', available)]
-  
+
+  # Find the files that match the request
   idx <- do.call(c, lapply(request, function(pattern) {
     grep(pattern = pattern, x = available)
   }))
-  
+
   if (any(idx < 1))
     return()
-  
-  unique(available[idx])
+
+  available[idx]  # Available files that match the request
```

Comment:
I seem to remember `idx` can potentially result in duplicate values, so `unique` was needed to avoid hymodelc blowing up.

Reply:
I moved `unique` to the end of the `request` pipe to reduce the number of `grep` calls. Im struggling to image a scenario where unique patterns would match the same file, unless `met_file_format` is poorly specified? Maybe worth calling `unique` twice though?

Reply:
Might not be an issue on the happy path. But unless you’re confident that it’ll won’t return dupes under all possible inputs, probably safer to explicitly dedupe the output 🤷‍♂️